File: CSharpCodeBase/components/controlsystem/actions/actionattackmouse.lua
local Action = require 'components.controlsystem.action'
local Class = require 'utils.hump.class'
local Vector3 = require 'utils.hump.vector3'
local HitTester = require 'utils.hittest'

local STATE_MOVING = 0
local STATE_REQUEST_ATTACK = 1
local STATE_ATTACK = 2

local ActionFastAttack = Class({__includes=Action, init = function (self, entity, target)
      Action.init(self, 2, false, 'fastattack')
      self.entity = entity
      self.target = target
      self.state = STATE_REQUEST_ATTACK
end})

function ActionFastAttack:OnPushed()
  self.priority = 3
end

function ActionFastAttack:OnStartRunning()
  print("Start attacking!")
  --self.priority = 3
  self.state = STATE_REQUEST_ATTACK
  self.entity.combat:AttackVector(self.target)
  self.entity.mover:SetSpeed(self.entity.config:Get('moverSpeed') / 4)
end

function ActionFastAttack:OnStopRunning()
  self.entity.mover:ResetSpeed()
  Action.OnStopRunning(self)
  --self.entity.mecanim:ForceSetState('Empty', 1)
end


function ActionFastAttack:Update()
  self:ProcessMoving()
  return self.state == STATE_ATTACK 
end

function ActionFastAttack:ProcessMoving()
  if GameController.inputService:LeftStickYIsPressed() or  GameController.inputService:LeftStickXIsPressed() then
    self.currentInput = GameController.inputService:LeftStickValues()
    self.currentInput:RotateAroundY(GameController.camera.angle) 
    self.entity.mover:SetInput(self.currentInput)
  end
end

function ActionFastAttack:OnEvent(data)
  if data == "punch" then  
    self.state = STATE_ATTACK
    -- Get damage objects
    local miss = true
    if self.entity.weaponContainer.currentWeapon.remote > 0 then
      -- shoot
      local sourceForward = Transform.GetForwardVector(self.entity.transform)
      self.entity.weaponContainer:Attack(self.entity:GetPosition() + self.target)
    else
      local nearEntities = RaycastUtils.GetEntitiesInRadius(self.entity:GetPosition(), 5)
      local sourceForward = Transform.GetForwardVector(self.entity.transform)
      for k, v in pairs(nearEntities) do
        if v ~= self.entity and v.interactable then
          if HitTester.CheckHitEntity(self.entity, sourceForward, v, 45, 1.5) then
            self.entity.damageProcessor:SendDamage(v, self.entity.weaponContainer.currentWeapon:GetDamage())
            if GameController.inputService:IsGamepad() and self.entity.gamepadRightStickController:GetTarget() == nil and v.gameObject.tag ~= 'Player' and v.gameObject.tag == 'Enemy' then
              self.entity.gamepadRightStickController:SetTarget(v)
            end
            miss = false
          end
        end
      end
    end
    if miss then
      self.entity.combat:ResetCombo()
    end
  end
end

function ActionFastAttack:IsContinuous()
  return false
end

return ActionFastAttack

File: CSharpCodeBase/components/controlsystem/actions/actionattackmouse.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class actionattackmouse :action{
 var STATE_MOVING = 0;
 var STATE_REQUEST_ATTACK = 1;
 var STATE_ATTACK = 2;
 var ActionFastAttack = Class({__includes=Action, init = public void (self, entity, target){
       Action.init(self, 2, false, "fastattack");
       this.entity = entity;
       this.target = target;
       this.state = STATE_REQUEST_ATTACK;
 }})
 public void ActionFastAttack:OnPushed(){
   this.priority = 3;
 }
 public void ActionFastAttack:OnStartRunning(){
   print("Start attacking!");
   //this.priority = 3
   this.state = STATE_REQUEST_ATTACK;
   this.entity.combat:AttackVector(this.target);
   this.entity.mover:SetSpeed(this.entity.config:Get("moverSpeed") / 4);
 }
 public void ActionFastAttack:OnStopRunning(){
   this.entity.mover:ResetSpeed();
   Action.OnStopRunning(self);
   //this.entity.mecanim:ForceSetState("Empty", 1)
 }
 public void ActionFastAttack:Update(){
   self:ProcessMoving();
   return this.state == STATE_ATTACK ;
 }
 public void ActionFastAttack:ProcessMoving(){
   if(GameController.inputService:LeftStickYIsPressed()  ||   GameController.inputService:LeftStickXIsPressed()  ){ 
     this.currentInput = GameController.inputService:LeftStickValues();
     this.currentInput:RotateAroundY(GameController.camera.angle) ;
     this.entity.mover:SetInput(this.currentInput);
   }
 }
 public void ActionFastAttack:OnEvent(data){
   if(data == "punch"  ){   
     this.state = STATE_ATTACK;
     // Get damage objects
     var miss = true;
     if(this.entity.weaponContainer.currentWeapon.remote > 0  ){ 
       // shoot
       var sourceForward = Transform.GetForwardVector(this.entity.transform);
       this.entity.weaponContainer:Attack(this.entity:GetPosition() + this.target);
     }else{
       var nearEntities = RaycastUtils.GetEntitiesInRadius(this.entity:GetPosition(), 5);
       var sourceForward = Transform.GetForwardVector(this.entity.transform);
       for(k, v in pairs(nearEntities) ){
         if(v != this.entity  &&  v.interactable  ){ 
           if(HitTester.CheckHitEntity(this.entity, sourceForward, v, 45, 1.5)  ){ 
             this.entity.damageProcessor:SendDamage(v, this.entity.weaponContainer.currentWeapon:GetDamage());
             if(GameController.inputService:IsGamepad()  &&  this.entity.gamepadRightStickController:GetTarget() == null  &&  v.gameObject.tag != "Player"  &&  v.gameObject.tag == "Enemy"  ){ 
               this.entity.gamepadRightStickController:SetTarget(v);
             }
             miss = false;
           }
         }
       }
     }
     if(miss  ){ 
       this.entity.combat:ResetCombo();
     }
   }
 }
 public void ActionFastAttack:IsContinuous(){
   return false;
 }
 return ActionFastAttack}}